src/core/smack.c

```diff
@@ -0,0 +1,953 @@
+    bs->bit_index = 0;
+    return bs;
+}
+
+static inline int read_bit(bitstream *bs)
+{
+    if (bs->index >= bs->length) {
+        return 0;
+    }
+    int result = bs->data[bs->index] & (1 << bs->bit_index);
+    if (++bs->bit_index >= 8) {
+        bs->index++;
+        bs->bit_index = 0;
+    }
+    return result ? 1 : 0;
+}
+
+static inline uint8_t read_byte(bitstream *bs)
+{
+    if (bs->bit_index == 0) {
+        // special case: on exact byte boundary
+        if (bs->index < bs->length) {
+            return bs->data[bs->index++];
+        } else {
+            return 0;
+        }
+    }
+    if (bs->index + 1 >= bs->length) {
+        return 0;
+    }
+    uint8_t value = bs->data[bs->index] >> bs->bit_index;
+    bs->index++;
+    value |= (bs->data[bs->index] & BIT_MASKS[bs->bit_index]) << (8 - bs->bit_index);
+    return value;
+}
+
+// 8-bit huffman tree functions
+
+static huffnode8 *build_tree8_nodes(bitstream *bs, hufftree8 *tree, int value, int bits)
+{
+    huffnode8 *node = &tree->nodes[tree->size++];
+    if (read_bit(bs)) {
+        node->is_leaf = 0;
+        node->b[0] = build_tree8_nodes(bs, tree, value, bits + 1);
+        node->b[1] = build_tree8_nodes(bs, tree, value | (1 << bits), bits + 1);
+    } else {
+        node->is_leaf = 1;
+        node->value = read_byte(bs);
+    }
+    return node;
+}
+
+static hufftree8 *create_tree8(bitstream *bs)
+{
+    if (read_bit(bs)) {
+        hufftree8 *tree = (hufftree8 *) smk_malloc(sizeof(hufftree8));
+        if (!tree) {
+            log_error("SMK: no memory for 8-bit tree", 0, 0);
+            return NULL;
+        }
+        build_tree8_nodes(bs, tree, 0, 0);
+        if (read_bit(bs) != 0) {
+            log_error("SMK: 8-bit tree not closed", 0, 0);
+            free(tree);
+            return NULL;
+        }
+        return tree;
+    } else {
+        log_info("SMK: WARN: no 8-bit tree found", 0, 0);
+        return NULL;
+    }
+}
+
+static void free_tree8(hufftree8 *tree)
+{
+    free(tree);
+}
+
+static uint8_t lookup_tree8(bitstream *bs, hufftree8 *tree)
+{
+    huffnode8 *node = &tree->nodes[0];
+    while (!node->is_leaf) {
+        node = node->b[read_bit(bs)];
+    }
+    return node->value;
+}
+
+// 16-bit huffman tree functions
+
+static void free_node16(huffnode16 *node)
+{
+    if (!node) {
+        return;
+    }
+    if (!node->is_leaf) {
+        free_node16(node->b[0]);
+        free_node16(node->b[1]);
+    }
+    free(node);
+}
+
+static void free_tree16(hufftree16 *tree)
+{
+    if (!tree) {
+        return;
+    }
+    for (int i = 0; i < 3; i++) {
+        if (!tree->escape_nodes[i]->is_leaf) {
+            // Free manually allocated node because it's not in the tree
+            free(tree->escape_nodes[i]);
+        }
+    }
+    free_node16(tree->root);
+    free_tree8(tree->low);
+    free_tree8(tree->high);
+    free(tree);
+}
+
+static huffnode16 *build_tree16_nodes(bitstream *bs, hufftree16 *tree, int value, int bits)
```

Comment:
Same here. `int value` and `int trees` are passed on recursively but never actually used.